README.md
# Intcode Interpreter

[![Hackage](https://img.shields.io/hackage/v/intcode.svg)](https://hackage.haskell.org/package/intcode) [![Build Status](https://secure.travis-ci.org/glguy/intcode.png?branch=master)](http://travis-ci.org/glguy/intcode)

Implementation of the Intcode virtual machine as defined by
[Advent of Code 2019](https://adventofcode.com/2019).

This implementation provides an efficient, pure implementation
of the interpreter and exposes multiple levels of abstraction
to make it easy to use in a variety of situations.

This implementation is derived from my puzzle solutions
[glguy/advent2019](https://github.com/glguy/advent2019).
Example uses include
[Day13.hs](https://github.com/glguy/advent2019/blob/master/execs/Day13.hs)
and
[Day15.hs](https://github.com/glguy/advent2019/blob/master/execs/Day15.hs).


CHANGELOG.md
# Revision history for intcode

## 0.4.0.0
* Add `Intcode.Step`

## 0.3.0.0
* Split machine-state and opcodes into separate modules.
* Add `Intcode.Parse`
* Add `Intcode.hRunIO`

## 0.2.0.0
* Rename 'intCodeToList' to 'intcodeToList'

## 0.1.0.0
* Initial release


src/Intcode/Step.hs
{-# Language Safe #-}
{-|
Module      : Intcode.Step
Description : Intcode small-step semantics
Copyright   : (c) Eric Mertens, 2019,2020
License     : ISC
Maintainer  : emertens@gmail.com

This module advances a 'Machine' by interpreting the opcode at the
current program counter.

-}
module Intcode.Step (Step(..), step) where

import Data.Traversable    (mapAccumL)
import Text.Show.Functions ()

import Intcode.Opcode      (Mode(..), Opcode(..), decode)
import Intcode.Machine     (Machine, (!), addRelBase, jmp, pc, relBase, set)

-- | Result of small-step semantics.
data Step
  = Step         !Machine    -- ^ update machine without output
  | StepOut !Int !Machine    -- ^ update machine with output
  | StepIn  (Int -> Machine) -- ^ machine blocked waiting for input
  | StepHalt                 -- ^ halt
  | StepFault                -- ^ bad instruction
  deriving Show

-- | Small-step semantics of virtual machine.
step :: Machine -> Step
step mach =
  case populateParams <$> decode (mach ! pc mach) of
    Nothing            -> StepFault
    Just (pc', opcode) -> opcodeImpl opcode $! jmp pc' mach

  where
    populateParams :: Opcode Mode -> (Int, Opcode Int)
    populateParams = mapWithIndex toPtr (pc mach + 1)

    toPtr :: Int -> Mode -> Int
    toPtr i Imm =        i
    toPtr i Abs = mach ! i
    toPtr i Rel = mach ! i + relBase mach

-- | Apply a decoded opcode to the machine state.
opcodeImpl ::
  Opcode Int {- ^ opcode with pointers    -} ->
  Machine    {- ^ machine with PC updated -} ->
  Step
opcodeImpl o m =
  case o of
    Add a b c -> Step    (set c (at a + at b) m)
    Mul a b c -> Step    (set c (at a * at b) m)
    Inp a     -> StepIn  (\i -> set a i m)
    Out a     -> StepOut (at a) m
    Jnz a b   -> Step    (if at a /= 0 then jmp (at b) m else m)
    Jz  a b   -> Step    (if at a == 0 then jmp (at b) m else m)
    Lt  a b c -> Step    (set c (if at a <  at b then 1 else 0) m)
    Eq  a b c -> Step    (set c (if at a == at b then 1 else 0) m)
    Arb a     -> Step    (addRelBase (at a) m)
    Hlt       -> StepHalt
  where
    at i = m ! i

mapWithIndex :: (Int -> a -> b) -> Int -> Opcode a -> (Int, Opcode b)
mapWithIndex f = mapAccumL (\i a -> (i+1, f i a))
{-# INLINE mapWithIndex #-}


src/Intcode/Parse.hs
{-# Language Safe #-}
{-|
Module      : Intcode.Parse
Description : Intcode source file parser
Copyright   : (c) Eric Mertens, 2019
License     : ISC
Maintainer  : emertens@gmail.com

This module implements a parser for the simple comma, separated format
used in the Advent of Code input files.

-}
module Intcode.Parse (parseInts) where

-- | Parse a list of comma separated integers.
--
-- >>> parseInts "1, - 2, 3,-4"
-- Just [1,-2,3,-4]
--
-- >>> parseInts " "
-- Just []
--
-- >>> parseInts "1,2,3,x"
-- Nothing
parseInts ::
  String      {- ^ parser input    -} ->
  Maybe [Int] {- ^ parsed integers -}
parseInts str
  | [(i,str1)] <- reads str = parseInts' [i] str1
  | [("","")]  <- lex str   = Just []
  | otherwise               = Nothing

-- | Helper function for 'parseInts'
parseInts' ::
  [Int]       {- ^ reversed accumulator -} ->
  String      {- ^ parser input         -} ->
  Maybe [Int] {- ^ parsed integers      -}
parseInts' xs str =
  case lex str of
    [(",",str1)] | [(x,str2)] <- reads str1 -> parseInts' (x:xs) str2
    [("","")]                               -> Just (reverse xs)
    _                                       -> Nothing


src/Intcode/Opcode.hs
{-# Language DeriveTraversable, Safe #-}
{-|
Module      : Intcode.Opcode
Description : Intcode opcodes
Copyright   : (c) Eric Mertens, 2019
License     : ISC
Maintainer  : emertens@gmail.com

This module provides a representation of the intcode machine's opcodes.

Opcodes are parameterized over their parameters. This allows the
implementation to store both parameter modes and resolved parameter
pointers in the same constructors.

-}
module Intcode.Opcode
  (
  -- * Types
  Opcode(..), Mode(..),

  -- * Decoder
  decode,
  ) where

------------------------------------------------------------------------
-- Opcode decoder
------------------------------------------------------------------------

-- | Parameter modes
data Mode
  = Abs -- ^ absolute position
  | Imm -- ^ immediate
  | Rel -- ^ relative position
  deriving (Eq, Ord, Read, Show)

-- | Opcodes parameterized over argument representations.
data Opcode a
  = Add !a !a !a -- ^ __addition:__        @c = a + b@
  | Mul !a !a !a -- ^ __multiplication:__  @c = a * b@
  | Inp !a       -- ^ __input:__           @a = input()@
  | Out !a       -- ^ __output:__          @output(a)@
  | Jnz !a !a    -- ^ __jump-if-true:__    @if a then goto b@
  | Jz  !a !a    -- ^ __jump-if-false:__   @if !a then goto b@
  | Lt  !a !a !a -- ^ __less-than:__       @c = a < b@
  | Eq  !a !a !a -- ^ __equals:__          @c = a == b@
  | Arb !a       -- ^ __adjust-rel-base:__ @rel += a@
  | Hlt          -- ^ __halt__
  deriving (Eq, Ord, Read, Show, Functor, Foldable)

-- | Decode an instruction to determine the opcode and parameter modes.
--
-- >>> decode 1002
-- Just (Mul Abs Imm Abs)
decode :: Int {- ^ opcode -} -> Maybe (Opcode Mode)
decode n =
  case n `rem` 100 of
    1  -> fill (Add 1 2 3)
    2  -> fill (Mul 1 2 3)
    3  -> fill (Inp 1    )
    4  -> fill (Out 1    )
    5  -> fill (Jnz 1 2  )
    6  -> fill (Jz  1 2  )
    7  -> fill (Lt  1 2 3)
    8  -> fill (Eq  1 2 3)
    9  -> fill (Arb 1    )
    99 -> fill Hlt
    _  -> Nothing
  where
    fill = traverse (parameter n)
{-# INLINABLE decode #-}

-- | Compute the parameter mode for an argument at a given position.
parameter ::
  Int {- ^ opcode   -} ->
  Int {- ^ position -} ->
  Maybe Mode
parameter n i =
  case digit (i+1) n of
    0 -> Just Abs
    1 -> Just Imm
    2 -> Just Rel
    _ -> Nothing

-- | Arguments visited from left to right.
instance Traversable Opcode where
  {-# INLINE traverse #-}
  traverse f o =
    case o of
      Add x y z -> Add <$> f x <*> f y <*> f z
      Mul x y z -> Mul <$> f x <*> f y <*> f z
      Inp x     -> Inp <$> f x
      Out x     -> Out <$> f x
      Jnz x y   -> Jnz <$> f x <*> f y
      Jz  x y   -> Jz  <$> f x <*> f y
      Lt  x y z -> Lt  <$> f x <*> f y <*> f z
      Eq  x y z -> Eq  <$> f x <*> f y <*> f z
      Arb x     -> Arb <$> f x
      Hlt       -> pure Hlt

-- | Extract the ith digit from a number.
--
-- >>> digit 0 2468
-- 8
-- >>> digit 3 2468
-- 2
-- >>> digit 4 2468
-- 0
digit :: Int {- ^ position -} -> Int {- ^ number -} -> Int {- ^ digit -}
digit i x = x `quot` (10^i) `rem` 10


src/Intcode/Machine.hs
{-# Language Trustworthy #-}
{-|
Module      : Intcode.Machine
Description : Intcode machine representation
Copyright   : (c) Eric Mertens, 2019
License     : ISC
Maintainer  : emertens@gmail.com

The module implements the representation of the intcode machine state.

The 'Machine' type stores the initial memory image in an array and
only stores changes to that initial image. This allows for more efficient
comparisons of machine states for equality when there are few changes to
memory.

This implementation of the machine supports negative memory addresses.
These are defined not to be used in the Advent of Code problems.

This implementation stores machine-sized 'Int' values in memory.

-}
module Intcode.Machine
  (
  -- * Machine state
  Machine(..), new,

  -- * Register operations
  jmp, addRelBase,

  -- * Memory operations
  (!), set, memoryList,
  )
 where

import           Data.IntMap (IntMap)
import qualified Data.IntMap as IntMap
import qualified Data.Primitive.PrimArray as P

-- | Machine state is comprised of the program counter, relative base
-- pointer, and memory.
--
-- * Interact with registers using: 'jmp', 'addRelBase'
-- * Interact with memory using: ('!'), 'set'
-- * Build new machines with: 'new'
--
-- Updates to memory are stored separately from the initial values
-- which can enable equality comparisons to be relatively efficient.
-- This efficiency comes from being able to compare the inital memory
-- using only pointer equality when two machines are created by the
-- same call to 'new'.
data Machine = Machine
  { pc         :: !Int          -- ^ program counter
  , relBase    :: !Int          -- ^ relative base pointer
  , memUpdates :: !(IntMap Int) -- ^ memory updates
  , memInitial :: {-# Unpack #-} !(P.PrimArray Int) -- ^ initial memory
  }
  deriving (Eq, Ord, Show)

-- | Value stored in initial memory image at given index.
indexImage ::
  Machine {- ^ machine  -} ->
  Int     {- ^ position -} ->
  Int     {- ^ value    -}
indexImage m i
  | i < P.sizeofPrimArray a, 0 <= i = P.indexPrimArray a i
  | otherwise                       = 0
  where
    a = memInitial m
{-# INLINE indexImage #-}

-- | Memory lookup.
(!) ::
  Machine {- ^ machine  -} ->
  Int     {- ^ position -} ->
  Int     {- ^ value    -}
m ! i = IntMap.findWithDefault (indexImage m i) i (memUpdates m)
{-# INLINE (!) #-}

-- | Construct machine from a list of initial values starting
-- at address 0. Program counter and relative base start at 0.
new ::
  [Int] {- ^ initial memory -} ->
  Machine
new initialValues = Machine
  { pc         = 0
  , relBase    = 0
  , memUpdates = IntMap.empty
  , memInitial = P.primArrayFromList initialValues
  }

-- | Store value at given memory position.
set ::
  Int {- ^ position -} ->
  Int {- ^ value    -} ->
  Machine -> Machine
set i v m
  | v == o    = m { memUpdates = IntMap.delete i   (memUpdates m) }
  | otherwise = m { memUpdates = IntMap.insert i v (memUpdates m) }
  where
    o = indexImage m i

-- | Add offset to relative base pointer.
addRelBase ::
  Int {- ^ offset -} ->
  Machine -> Machine
addRelBase i mach = mach { relBase = relBase mach + i }
{-# INLINE addRelBase #-}

-- | Set program counter to a new address.
jmp ::
  Int {- ^ program counter -} ->
  Machine -> Machine
jmp i mach = mach { pc = i }
{-# INLINE jmp #-}

-- | Generate a list representation of memory starting from
-- zero. This can get big for sparsely filled memory using
-- large addresses. Returned values start at position 0.
--
-- >>> memoryList (set 8 10 (new [1,2,3]))
-- [1,2,3,0,0,0,0,0,10]
memoryList ::
  Machine ->
  [Int] {- ^ memory values -}
memoryList mach
  | IntMap.null (memUpdates mach) = P.primArrayToList (memInitial mach)
  | otherwise                  = [mach ! i | i <- [0 .. top]]
  where
    top = max (P.sizeofPrimArray (memInitial mach) - 1)
              (fst (IntMap.findMax (memUpdates mach)))


src/Intcode.hs
{-# Language Safe #-}
{-|
Module      : Intcode
Description : Intcode interpreter
Copyright   : (c) Eric Mertens, 2019
License     : ISC
Maintainer  : emertens@gmail.com

Intcode is a virtual machine environment defined to have some arithmetic,
conditional jumps, and simple input and output facilities.

The instruction set is designed with independently selectable address modes for
each of its input and output parameters. The architecture is designed to be
simple to implement while powerful enough to write interesting programs
efficiently. The addition of a /relative base pointer/ makes it easy to
implement function calls in the language.

This Intcode architecture is defined across multiple
<https://adventofcode.com/2019/about Advent of Code 2019> tasks:
<https://adventofcode.com/2019/day/2 2>,
<https://adventofcode.com/2019/day/5 5>,
<https://adventofcode.com/2019/day/7 7>, and
<https://adventofcode.com/2019/day/9 9>

Common use modes:

* Machine construction: 'new'
* List functions: 'intcodeToList', 'effectList'
* Individual machine step processing: 'Step', 'step'
* Input/output interpretation: 'Effect', 'run'

Submodules:

* "Intcode.Machine" exposes the implementation details of the interpreter state.
* "Intcode.Parse" provides a parser for intcode text files.
* "Intcode.Opcode" provides types and the decoder for opcodes.

-}
module Intcode
  (
  -- * Simple list interface
  intcodeToList,

  -- * Machine state
  Machine, (!), new, set, memoryList,

  -- * Big-step semantics
  Effect(..), run,

  -- * Effect operations
  effectList,

  -- * Small-step semantics
  Step(..), step,

  -- * Exceptions
  IntcodeFault(..),

  -- * ASCII I/O interface
  runIO, hRunIO,

  ) where

import Control.Exception   (Exception(..), throw, throwIO)
import Data.Char           (chr, ord)
import System.IO           (Handle, hGetChar, hPutChar, hPutStrLn, stdin, stdout)

import Intcode.Machine     (Machine(..), (!), memoryList, new, set)
import Intcode.Step        (Step(..), step)

------------------------------------------------------------------------
-- ASCII I/O
------------------------------------------------------------------------

-- | Run intcode program using stdio. Non-ASCII outputs are printed as
-- integers.
--
-- Note that input and output is affected by handle buffering modes.
--
-- >>> runIO (run (new [104,72,104,101,104,108,104,108,104,111,104,33,104,10,99]))
-- Hello!
--
-- >>> runIO (run (new [104,-50,104,1000,99]))
-- <<-50>>
-- <<1000>>
runIO :: Effect -> IO ()
runIO = hRunIO stdin stdout

-- | 'runIO' generalized to an arbitrary input and output handle.
hRunIO ::
  Handle {- ^ input handle  -} ->
  Handle {- ^ output handle -} ->
  Effect {- ^ effect        -} ->
  IO ()
hRunIO inH outH = go
  where
    go (Output o e)
      | 0 <= o, o < 0x80 = hPutChar outH (chr (fromIntegral o)) >> go e
      | otherwise        = hPutStrLn outH ("<<" ++ show o ++ ">>") >> go e
    go (Input f)         = go . f . fromIntegral . ord =<< hGetChar inH
    go Halt              = return ()
    go Fault             = throwIO IntcodeFault

------------------------------------------------------------------------
-- High-level interface
------------------------------------------------------------------------

-- | Run a given memory image as a list transducer.
--
-- Use 'effectList' when you want to provide a specific 'Effect'.
--
-- Throws: 'IntcodeFault' when machine faults or too few inputs are provided.
--
--
-- >>> intcodeToList [3,12,6,12,15,1,13,14,13,4,13,99,-1,0,1,9] <$> [[0],[10]]
-- [[0],[1]]
--
-- >>> intcodeToList [3,3,1105,-1,9,1101,0,0,12,4,12,99,1] <$> [[0],[10]]
-- [[0],[1]]
--
-- >>> :{
-- >>> intcodeToList
-- >>>   [3,21,1008,21,8,20,1005,20,22,107,8,21,20,1006,20,31,
-- >>>    1106,0,36,98,0,0,1002,21,125,20,4,20,1105,1,46,104,
-- >>>    999,1105,1,46,1101,1000,1,20,4,20,1105,1,46,98,99]
-- >>> <$> [[7],[8],[9]]
-- >>> :}
-- [[999],[1000],[1001]]
intcodeToList ::
  [Int] {- ^ initial memory -} ->
  [Int] {- ^ inputs         -} ->
  [Int] {- ^ outputs        -}
intcodeToList = effectList . run . new

-- | Evaluate a program's effect as a function from a list of
-- inputs to a list of outputs.
--
-- Throws: 'IntcodeFault' when machine faults or too few inputs are provided.
effectList ::
  Effect {- ^ program effect -} ->
  [Int]  {- ^ inputs         -} ->
  [Int]  {- ^ outputs        -}
effectList effect inputs =
  case effect of
    Fault      -> throw IntcodeFault
    Halt       -> []
    Output o e -> o : effectList e inputs
    Input f    ->
      case inputs of
        x:xs -> effectList (f x) xs
        []   -> throw IntcodeFault

------------------------------------------------------------------------
-- Big-step semantics
------------------------------------------------------------------------

-- | Possible effects from running a machine
data Effect
  = Output !Int Effect    -- ^ Output an integer
  | Input (Int -> Effect) -- ^ Input an integer
  | Halt                  -- ^ Halt execution
  | Fault                 -- ^ Execution failure
  deriving Show

-- | Big-step semantics of virtual machine. The implementation details
-- of 'Machine' are abstracted away and the program behavior can be
-- observed by interpreting the various 'Effect' constructors.
--
-- >>> run (new [1102,34915192,34915192,7,4,7,99,0])
-- Output 1219070632396864 Halt
--
-- >>> run (new [3,1,99])
-- Input <function>
run :: Machine -> Effect
run mach =
  case step mach of
    Step mach'        -> run mach'
    StepOut out mach' -> Output out (run mach')
    StepIn f          -> Input (run . f)
    StepHalt          -> Halt
    StepFault         -> Fault

------------------------------------------------------------------------
-- Exceptions
------------------------------------------------------------------------

-- | Error when a machine fails to decode an instruction.
data IntcodeFault = IntcodeFault
  deriving (Eq, Ord, Show, Read)

instance Exception IntcodeFault where
  displayException _ = "intcode machine fault"
